File: howdy/src/compare.py
# Compare incoming video with known faces
# Running in a local python instance to get around PATH issues

# Import time so we can start timing asap
import time

# Start timing
timings = {"st": time.time()}

# Import required modules
import configparser
import json
import os
import sys
import threading
from datetime import datetime, timezone
from typing import cast

import cv2
import numpy as np
import paths_factory
import snapshot
from core.detector import BACKEND_NAME, FaceModel, clahe_enabled, create_clahe
from i18n import _
from recorders.video_capture import VideoCapture


def exit(code=None):
    if code is not None:
        sys.exit(code)


def init_detector(lock):
    global face_model
    face_model = FaceModel(config)
    timings["ll"] = time.time() - timings["ll"]
    lock.release()


def make_snapshot(
    type,
    snapframes_list: list,
    frames_count: int,
    best_score_val: float,
    frame_time: float,
):
    """Generate snapshot after detection"""
    snapshot.generate(
        snapframes_list,
        [
            type + _(" LOGIN"),
            _("Date: ") + datetime.now(timezone.utc).strftime("%Y/%m/%d %H:%M:%S UTC"),
            _("Scan time: ") + str(round(frame_time, 2)) + "s",
            _("Frames: ")
            + str(frames_count)
            + " ("
            + str(round(frames_count / frame_time, 2))
            + "FPS)",
            _("Hostname: ") + os.uname().nodename,
            _("Best match score: ") + str(round(best_score_val, 3)),
        ],
    )


def update_best_score(current: float | None, score: float, metric: str) -> float:
    if current is None:
        return score
    if metric == "cosine":
        return max(current, score)
    return min(current, score)


# Make sure we were given an username to test against
if len(sys.argv) < 2:
    exit(12)

# The username of the user being authenticated
user = sys.argv[1]
# The model file contents
models = []
# Encoded face models
encodings = []
# Model metadata per encoding
encoding_models = []
# Amount of ignored 100% black frames
black_tries = 0
# Amount of ignored dark frames
dark_tries = 0
# Total amount of frames captured
frames = 0
# Captured frames for snapshot capture
snapframes = []
# Tracks the best score in the loop
best_score = None
face_model: FaceModel | None = None

# Try to load the face model from the models folder
try:
    with open(paths_factory.user_model_path(user)) as f:
        models = json.load(f)
    for model in models:
        if model.get("backend") != BACKEND_NAME:
            print(
                _("Stored face models use an incompatible backend; re-enroll required")
            )
            exit(10)
        for encoding in model["data"]:
            encodings.append(encoding)
            encoding_models.append(model)
except FileNotFoundError:
    exit(10)

# Check if the file contains a model
if len(models) < 1 or len(encodings) < 1:
    exit(10)

# Read config from disk
config = configparser.ConfigParser()
config.read(paths_factory.config_file_path())

# Get all config values needed
timeout = config.getint("video", "timeout", fallback=4)
dark_threshold = config.getfloat("video", "dark_threshold", fallback=50.0)
end_report = config.getboolean("debug", "end_report", fallback=False)
save_failed = config.getboolean("snapshots", "save_failed", fallback=False)
save_successful = config.getboolean("snapshots", "save_successful", fallback=False)
rotate = config.getint("video", "rotate", fallback=0)
timings["in"] = time.time() - timings["st"]

# Import face recognition, takes some time
timings["ll"] = time.time()

# Start threading and wait for init to finish
lock = threading.Lock()
lock.acquire()
threading.Thread(target=init_detector, args=(lock,), daemon=True).start()

# Start video capture on the IR camera
timings["ic"] = time.time()

video_capture = VideoCapture(config)

# Read exposure from config to use in the main loop
exposure = config.getint("video", "exposure", fallback=-1)

# Note the time it took to open the camera
timings["ic"] = time.time() - timings["ic"]

# wait for thread to finish
lock.acquire()
lock.release()
del lock

if face_model is None:
    print(_("Face model not initialized"))
    exit(1)
assert face_model is not None

active_face_model = cast(FaceModel, face_model)
known_encodings = np.asarray(encodings, dtype=np.float32)

# Fetch the max frame height
max_height = config.getfloat("video", "max_height", fallback=320.0)

# Get the height of the image (which would be the width if screen is portrait oriented)
height = video_capture.internal.get(cv2.CAP_PROP_FRAME_HEIGHT) or 1
if rotate == 2:
    height = video_capture.internal.get(cv2.CAP_PROP_FRAME_WIDTH) or 1
# Calculate the amount the image has to shrink
scaling_factor = max_height / max(height, 1)

# Initiate histogram equalization
clahe = create_clahe(config)

# Start the read loop
valid_frames = 0
timings["fr"] = time.time()
dark_running_total = 0

while True:
    # Increment the frame count every loop
    frames += 1

    # Stop if we've exceeded the time limit
    if time.time() - timings["fr"] > timeout:
        # Create a timeout snapshot if enabled
        if save_failed:
            make_snapshot(
                _("FAILED"),
                snapframes,
                frames,
                best_score or 0.0,
                time.time() - timings["fr"],
            )

        if dark_tries > 0 and valid_frames == 0:
            print(_("All frames were too dark, please check dark_threshold in config"))
            print(
                _("Average darkness: {avg}, Threshold: {threshold}").format(
                    avg=str(dark_running_total / max(1, valid_frames)),
                    threshold=str(dark_threshold),
                )
            )
            exit(13)
        else:
            exit(11)

    # Grab a single frame of video
    frame, gsframe = video_capture.read_frame()
    if gsframe.ndim != 2:
        gsframe = cv2.cvtColor(gsframe, cv2.COLOR_BGR2GRAY)
    if clahe_enabled(config):
        gsframe = clahe.apply(gsframe)

    # Create a histogram of the image with 8 values
    hist = cv2.calcHist([gsframe], [0], None, [8], [0, 256])
    # All values combined for percentage calculation
    hist_total = np.sum(hist)

    # If the image is fully black due to a bad camera read,
    # skip to the next frame
    if hist_total == 0:
        black_tries += 1
        continue

    # Calculate frame darkness
    darkness = hist[0] / hist_total * 100

    if darkness == 100:
        black_tries += 1
        continue

    dark_running_total += darkness
    valid_frames += 1

    # If the image exceeds darkness threshold due to subject distance,
    # skip to the next frame
    if darkness > dark_threshold:
        dark_tries += 1
        continue

    # If the height is too high
    if scaling_factor != 1:
        # Apply that factor to the frame
        gsframe = cv2.resize(
            gsframe,
            None,
            fx=scaling_factor,
            fy=scaling_factor,
            interpolation=cv2.INTER_AREA,
        )

    # If camera is configured to rotate = 1, check portrait in addition to landscape
    if rotate == 1:
        if frames % 3 == 1:
            gsframe = cv2.rotate(gsframe, cv2.ROTATE_90_COUNTERCLOCKWISE)
        if frames % 3 == 2:
            gsframe = cv2.rotate(gsframe, cv2.ROTATE_90_CLOCKWISE)

    # If camera is configured to rotate = 2, check portrait orientation
    elif rotate == 2:
        if frames % 2 == 0:
            gsframe = cv2.rotate(gsframe, cv2.ROTATE_90_COUNTERCLOCKWISE)
        else:
            gsframe = cv2.rotate(gsframe, cv2.ROTATE_90_CLOCKWISE)

    frame = active_face_model.prepare_frame(gsframe)

    # If snapshots have been turned on
    if save_failed or save_successful:
        # Start capturing frames for the snapshot
        if len(snapframes) < 3:
            snapframes.append(frame)

    # Get all faces from that frame as encodings
    face_locations = active_face_model.detect(frame)
    for face in face_locations:
        face_encoding = active_face_model.encode(frame, face)
        match = active_face_model.best_match(known_encodings, face_encoding)
        best_score = update_best_score(
            best_score, match.score, active_face_model.metric
        )

        # Check if a match that's confident enough
        if match.accepted:
            timings["tt"] = time.time() - timings["st"]
            timings["fl"] = time.time() - timings["fr"]

            # If set to true in the config, print debug text
            if end_report:

                def print_timing(label, k):
                    """Helper function to print a timing from the list"""
                    print("  %s: %dms" % (label, round(timings[k] * 1000)))

                # Print a nice timing report
                print(_("Time spent"))
                print_timing(_("Starting up"), "in")
                print(
                    _("  Open cam + load libs: %dms")
                    % (
                        round(
                            max(timings["ll"], timings["ic"]) * 1000,
                        )
                    )
                )
                print_timing(_("  Opening the camera"), "ic")
                print_timing(_("  Importing recognition libs"), "ll")
                print_timing(_("Searching for known face"), "fl")
                print_timing(_("Total time"), "tt")

                print(_("\nResolution"))
                width = video_capture.fw or 1
                print(_("  Native: %dx%d") % (height, width))
                # Save the new size for diagnostics
                scale_height, scale_width = frame.shape[:2]
                print(_("  Used: %dx%d") % (scale_height, scale_width))

                # Show the total number of frames and calculate the FPS by dividing it by the total scan time
                print(
                    _("\nFrames searched: %d (%.2f fps)")
                    % (frames, frames / timings["fl"])
                )
                print(_("Black frames ignored: %d ") % (black_tries,))
                print(_("Dark frames ignored: %d ") % (dark_tries,))
                print(_("Winning score: %.3f") % (match.score,))

                winning_model = encoding_models[match.index]
                print(
                    _('Winning model: %d ("%s")')
                    % (winning_model["id"], winning_model["label"])
                )

            # Make snapshot if enabled
            if save_successful:
                make_snapshot(
                    _("SUCCESSFUL"),
                    snapframes,
                    frames,
                    best_score or 0.0,
                    time.time() - timings["fr"],
                )

            # End peacefully
            exit(0)

    if exposure != -1:
        # For a strange reason on some cameras (e.g. Lenoxo X1E) setting manual exposure works only after a couple frames
        # are captured and even after a delay it does not always work. Setting exposure at every frame is reliable though.
        video_capture.internal.set(cv2.CAP_PROP_AUTO_EXPOSURE, 1.0)  # 1 = Manual
        video_capture.internal.set(cv2.CAP_PROP_EXPOSURE, float(exposure))

File: howdy/src/native/compare.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <limits>
#include <string>
#include <string_view>

#include <opencv2/imgproc.hpp>

#include "howdy/common/compare_exit.hpp"
#include "howdy/config/config_reader.hpp"
#include "howdy/config/runtime_paths.hpp"
#include "howdy/core/face_model.hpp"
#include "howdy/recorders/video_capture.hpp"
#include "howdy/storage/user_models.hpp"

namespace {

using howdy::native::CompareExit;

struct CompareArgs {
  std::string user;
  std::string config_path = howdy::native::resolve_config_path().string();
};

auto parse_args(int argc, char *argv[]) -> CompareArgs {
  CompareArgs args;

  for (int index = 1; index < argc; ++index) {
    const std::string_view arg(argv[index]);
    if (arg == "--help" || arg == "-h") {
      std::cout << "Usage: " << argv[0] << " [--config PATH] <user>\n";
      std::exit(static_cast<int>(CompareExit::kSuccess));
    }
    if (arg == "--config" && index + 1 < argc) {
      args.config_path = argv[++index];
      continue;
    }
    if (arg.starts_with('-')) {
      std::cerr << "Unknown argument: " << arg << "\n";
      std::exit(static_cast<int>(CompareExit::kAbort));
    }
    args.user = argv[index];
  }

  if (args.user.empty()) {
    std::exit(static_cast<int>(CompareExit::kAbort));
  }

  return args;
}

auto update_best_score(float current, float score, const std::string &metric)
    -> float {
  if (std::isnan(current)) {
    return score;
  }
  if (metric == "cosine") {
    return std::max(current, score);
  }
  return std::min(current, score);
}

auto apply_rotation(const cv::Mat &frame, int rotate, int frames) -> cv::Mat {
  if (rotate == 1) {
    if (frames % 3 == 1) {
      cv::Mat rotated;
      cv::rotate(frame, rotated, cv::ROTATE_90_COUNTERCLOCKWISE);
      return rotated;
    }
    if (frames % 3 == 2) {
      cv::Mat rotated;
      cv::rotate(frame, rotated, cv::ROTATE_90_CLOCKWISE);
      return rotated;
    }
  } else if (rotate == 2) {
    cv::Mat rotated;
    cv::rotate(frame, rotated,
               frames % 2 == 0 ? cv::ROTATE_90_COUNTERCLOCKWISE
                               : cv::ROTATE_90_CLOCKWISE);
    return rotated;
  }

  return frame;
}

}  // namespace

auto main(int argc, char *argv[]) -> int {
  const auto start_time = std::chrono::steady_clock::now();
  const CompareArgs args = parse_args(argc, argv);

  const auto loaded_models =
      howdy::native::load_user_models(args.user, howdy::native::FaceModel::kBackendName);
  if (loaded_models.status == howdy::native::UserModelStatus::kIncompatibleBackend) {
    std::cerr << loaded_models.error_message << "\n";
    return static_cast<int>(CompareExit::kNoFaceModel);
  }
  if (loaded_models.status == howdy::native::UserModelStatus::kParseError) {
    std::cerr << loaded_models.error_message << "\n";
    return static_cast<int>(CompareExit::kAbort);
  }
  if (loaded_models.status != howdy::native::UserModelStatus::kOk) {
    return static_cast<int>(CompareExit::kNoFaceModel);
  }

  howdy::native::ConfigReader config(args.config_path);
  if (!config.ok()) {
    std::cerr << "Failed to parse config: " << args.config_path << "\n";
    return static_cast<int>(CompareExit::kAbort);
  }

  howdy::native::FaceModel face_model(config);
  if (!face_model.ok()) {
    std::cerr << face_model.error_message() << "\n";
    return static_cast<int>(CompareExit::kAbort);
  }

  howdy::native::VideoCapture capture(howdy::native::load_capture_settings(config));
  if (!capture.open()) {
    std::cerr << capture.error_message() << "\n";
    return static_cast<int>(CompareExit::kInvalidDevice);
  }

  const int timeout = config.get_int("video", "timeout", 4);
  const float dark_threshold = config.get_float("video", "dark_threshold", 50.0F);
  const float max_height = config.get_float("video", "max_height", 320.0F);
  const int rotate = config.get_int("video", "rotate", 0);
  const int exposure = config.get_int("video", "exposure", -1);
  const bool end_report = config.get_bool("debug", "end_report", false);
  const bool use_clahe = config.get_bool("video", "clahe_enabled", true);
  const double clip_limit = config.get_float("video", "clahe_clip_limit", 1.25F);
  const int tile_size = config.get_int("video", "clahe_tile_grid_size", 8);

  auto native_height = capture.get(cv::CAP_PROP_FRAME_HEIGHT);
  if (rotate == 2) {
    native_height = capture.get(cv::CAP_PROP_FRAME_WIDTH);
  }
  const double scaling_factor = max_height / std::max(native_height, 1.0);

  cv::Ptr<cv::CLAHE> clahe;
  if (use_clahe) {
    clahe = cv::createCLAHE(clip_limit, cv::Size(tile_size, tile_size));
  }

  int frames = 0;
  int black_tries = 0;
  int dark_tries = 0;
  int valid_frames = 0;
  double dark_running_total = 0.0;
  float best_score = std::numeric_limits<float>::quiet_NaN();
  float winning_score = 0.0F;
  int winning_index = -1;
  auto frame_loop_start = std::chrono::steady_clock::now();

  while (true) {
    frames++;

    const auto elapsed =
        std::chrono::duration_cast<std::chrono::seconds>(
            std::chrono::steady_clock::now() - frame_loop_start)
            .count();
    if (elapsed > timeout) {
      if (dark_tries > 0 && valid_frames == dark_tries) {
        std::cerr << "All frames were too dark, please check dark_threshold in config\n";
        std::cerr << "Average darkness: "
                  << (dark_running_total / std::max(valid_frames, 1))
                  << ", Threshold: " << dark_threshold << "\n";
        return static_cast<int>(CompareExit::kTooDark);
      }
      return static_cast<int>(CompareExit::kTimeoutReached);
    }

    cv::Mat frame;
    cv::Mat gray_frame;
    if (!capture.read(frame, &gray_frame)) {
      std::cerr << capture.error_message() << "\n";
      return static_cast<int>(CompareExit::kInvalidDevice);
    }

    if (use_clahe) {
      clahe->apply(gray_frame, gray_frame);
    }

    cv::Mat hist;
    constexpr int hist_size[] = {8};
    constexpr float hist_range[] = {0.0F, 256.0F};
    const float *ranges[] = {hist_range};
    constexpr int channels[] = {0};
    cv::calcHist(&gray_frame, 1, channels, cv::Mat(), hist, 1, hist_size, ranges);
    const double hist_total = cv::sum(hist)[0];
    if (hist_total == 0.0) {
      black_tries++;
      continue;
    }

    const float darkness =
        static_cast<float>(hist.at<float>(0) / hist_total * 100.0);
    if (darkness == 100.0F) {
      black_tries++;
      continue;
    }

    dark_running_total += darkness;
    valid_frames++;

    if (darkness > dark_threshold) {
      dark_tries++;
      continue;
    }

    cv::Mat working_frame = gray_frame;
    if (scaling_factor != 1.0) {
      cv::resize(gray_frame, working_frame, cv::Size(), scaling_factor,
                 scaling_factor, cv::INTER_AREA);
    }
    working_frame = apply_rotation(working_frame, rotate, frames);

    cv::Mat prepared = face_model.prepare_frame(working_frame);
    const auto faces = face_model.detect(prepared);
    for (const auto &face : faces) {
      const auto encoding = face_model.encode(prepared, face);
      const auto match = face_model.best_match(loaded_models.stored.encodings, encoding);
      best_score = update_best_score(best_score, match.score, face_model.metric());

      if (!match.accepted) {
        continue;
      }

      winning_index = match.index;
      winning_score = match.score;

      if (end_report) {
        const auto total_ms =
            std::chrono::duration_cast<std::chrono::milliseconds>(
                std::chrono::steady_clock::now() - start_time)
                .count();
        std::cout << "Total time: " << total_ms << "ms\n";
        std::cout << "Frames searched: " << frames << "\n";
        std::cout << "Black frames ignored: " << black_tries << "\n";
        std::cout << "Dark frames ignored: " << dark_tries << "\n";
        std::cout << "Winning score: " << winning_score << "\n";
        if (winning_index >= 0 &&
            winning_index < static_cast<int>(loaded_models.stored.models.size())) {
          const auto &winner =
              loaded_models.stored.models[static_cast<std::size_t>(winning_index)];
          std::cout << "Winning model: " << winner.id << " (\"" << winner.label
                    << "\")\n";
        }
      }

      return static_cast<int>(CompareExit::kSuccess);
    }

    if (exposure != -1) {
      capture.set(cv::CAP_PROP_AUTO_EXPOSURE, 1.0);
      capture.set(cv::CAP_PROP_EXPOSURE, static_cast<double>(exposure));
    }
  }
}
